### PR TITLE
Implementation of the shadow Thing feature on the device (WIP)

### PR DESCRIPTION
**Pull request by ilcato**
Device implementation of shadow Thing feature.

The ArduinoIoTCloud update accept two new optional parameters:

void update(int mode = PROPERTIES_SYNC_FORCE_DEVICE, void (*callback)(void) = NULL);

the mode maybe:
  PROPERTIES_SYNC_FORCE_DEVICE, // default mode, after a connection/reconnection the device value of a property is forced on the cloud
  PROPERTIES_SYNC_FORCE_CLOUD, // after a connection/reconnection the cloud value of a property is forced on the device
  PROPERTIES_SYNC_AUTO, // after a connection/reconnection the timestamps of the changes on device and on cloud are used to choose which value the property will assume
  PROPERTIES_SYNC_CUSTOM // not yet implemented 

The callback is called after the processing.

### CODE REVIEW

**Comment by lxrobotics** [comment]
Could you please provide unit tests for your changes?

**Comment by lxrobotics** [comment]
Hi @ilcato,

I'm not really sure about your pull request :wink: especially the function that you've provided unit tests for: `getLastValues`. The function always encodes the exact same message (`{"n": "r:m", "vs": "getLastValues"}`and `{0: "r:m", 3: "getLastValues"}` respectively). This doesn't make sense to me. If the function should always return the same data we might as well store the resulting byte array hard coded and do not need to bother with encoding via CBOR library functions.

Another thing: The enum's `PROPERTIES_SYNC_CUSTOM` and `CLOUDSERIAL_SYNC` are defined but nowhere used. If they are not implemented they should not be in the code.

Lastly: Unit tests should exercise as much code as possible (in an ideal world we would have 100 % code coverage, branch coverage, ...) and the really juicy bits of the code are the many branches for the different modes. The unit tests for the function `getLastValues` which always returns the same data anyway does not provide much value.

**Comment by ilcato** [comment]
> Hi @ilcato,
> 
> I'm not really sure about your pull request 😉 especially the function that you've provided unit tests for: `getLastValues`. The function always encodes the exact same message (`{"n": "r:m", "vs": "getLastValues"}`and `{0: "r:m", 3: "getLastValues"}` respectively). This doesn't make sense to me. If the function should always return the same data we might as well store the resulting byte array hard coded and do not need to bother with encoding via CBOR library functions.
> 
> Another thing: The enum's `PROPERTIES_SYNC_CUSTOM` and `CLOUDSERIAL_SYNC` are defined but nowhere used. If they are not implemented they should not be in the code.
> 
> Lastly: Unit tests should exercise as much code as possible (in an ideal world we would have 100 % code coverage, branch coverage, ...) and the really juicy bits of the code are the many branches for the different modes. The unit tests for the function `getLastValues` which always returns the same data anyway does not provide much value.
> 
> **EDIT** You're not locked in on testing only through the interfaces of `ArduinoCloudThing` - testing of `ArduinoCloudProperty` is also an option.

The current implementation of the getLastValues method return a constant but the option to pass parameters was envisioned in order to optimize the number of properties returned via the broker; if such an option is not useful we can collapse it to a constant byte array.

CLOUDSERIAL_SYNC is used in the CloudSerial class and ArduinoIoTCloud in order to prevent the update method to break the sync functionality.

PROPERTIES_SYNC_CUSTOM is the a further sync mode that is not currently implemented.

The current proposed tests are only a trial in order to understand the whole unit test framework.

**Comment by lxrobotics** [comment]
Thank you for the clarification ;) I now understand that this PR is still WIP and will mark it as such.

File: src/ArduinoCloudThing.cpp [hunk 1]
```diff
@@ -475,31 +530,99 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+          int_property->writeByCloud(static_cast<int>(map_data->val.get())); /* Val is internally stored as float */
+          if(int_property->isAfterLastChange(cloudChangeEventTime)) {
+            int_property->forceCallbackOnChange();
+          }
+          int_property->setLastChangeTime(cloudChangeEventTime);
+        }
+        if(_mode == PROPERTIES_SYNC_AUTO){
+          if(int_property->isAfterLastChange(cloudChangeEventTime)) {
+           int_property->writeByCloud(static_cast<int>(map_data->val.get())); /* Val is internally stored as float */
+           int_property->forceCallbackOnChange();
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
            int_property->forceCallbackOnChange();
```

File: src/ArduinoCloudThing.cpp [hunk 2]
```diff
@@ -475,31 +530,99 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+        if(_mode == PROPERTIES_SYNC_FORCE_CLOUD){
+          float_property->writeByCloud(map_data->val.get());
+          if(float_property->isAfterLastChange(cloudChangeEventTime)) {
+            float_property->forceCallbackOnChange();
+          }
+          float_property->setLastChangeTime(cloudChangeEventTime);
+        }
+        if(_mode == PROPERTIES_SYNC_AUTO){
+          if(float_property->isAfterLastChange(cloudChangeEventTime)) {
+           float_property->writeByCloud(map_data->val.get());
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
            float_property->writeByCloud(map_data->val.get());
```

File: src/ArduinoCloudThing.cpp [hunk 3]
```diff
@@ -475,31 +530,99 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+           float_property->forceCallbackOnChange();
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
            float_property->forceCallbackOnChange();
```

File: src/ArduinoCloudThing.cpp [hunk 4]
```diff
@@ -475,31 +530,99 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+        if(_mode == PROPERTIES_SYNC_FORCE_CLOUD){
+          string_property->writeByCloud(map_data->str_val.get());
+          if(string_property->isAfterLastChange(cloudChangeEventTime)) {
+            string_property->forceCallbackOnChange();
+          }
+          string_property->setLastChangeTime(cloudChangeEventTime);
+        }
+        if(_mode == PROPERTIES_SYNC_AUTO){
+          if(string_property->isAfterLastChange(cloudChangeEventTime)) {
+           string_property->writeByCloud(map_data->str_val.get());
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
            string_property->writeByCloud(map_data->str_val.get());
```

File: src/ArduinoCloudThing.cpp [hunk 5]
```diff
@@ -475,31 +530,99 @@ ArduinoCloudThing::MapParserState ArduinoCloudThing::handle_LeaveMap(CborValue *
...
+           string_property->forceCallbackOnChange();
```
Comment (per1234) [inline src/ArduinoCloudThing.cpp]:
```suggestion
            string_property->forceCallbackOnChange();
```

**Pull request closed without merging**